From fff60836f2907dfa06a6a0669d00096d60afd778 Mon Sep 17 00:00:00 2001
From: Justin Suess <utilityemal77@gmail.com>
Date: Mon, 29 Jun 2026 17:38:06 -0400
Subject: [PATCH 09/13] arm64: dts: allwinner: a523: add EMAC 25 MHz reference
 clocks

The GMAC clock controller can drive the external RGMII PHY's reference
from an internal 25 MHz output. Add the EMAC0_25M/EMAC1_25M clocks (as
"emac25m") to gmac0/gmac1 and the matching PH13/PJ10 pinmux nodes, so a
board can select this path with allwinner,use-25m-clk instead of an
on-board crystal.
---
 .../arm64/boot/dts/allwinner/sun55i-a523.dtsi | 25 ++++++++++++++++---
 1 file changed, 21 insertions(+), 4 deletions(-)

diff --git a/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi b/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi
index a788faef07d6..2986d43cd4c3 100644
--- a/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun55i-a523.dtsi
@@ -422,6 +422,14 @@ rgmii0_pins: rgmii0-pins {
 				bias-disable;
 			};
 
+			rgmii0_25m_pin: rgmii0-25m-pin {
+				pins = "PH13";
+				allwinner,pinmux = <5>;
+				function = "gmac0";
+				drive-strength = <40>;
+				bias-disable;
+			};
+
 			rgmii1_pins: rgmii1-pins {
 				pins = "PJ0", "PJ1", "PJ2", "PJ3", "PJ4",
 				       "PJ5", "PJ6", "PJ7", "PJ8", "PJ9",
@@ -432,6 +440,14 @@ rgmii1_pins: rgmii1-pins {
 				bias-disable;
 			};
 
+			rgmii1_25m_pin: rgmii1-25m-pin {
+				pins = "PJ10";
+				allwinner,pinmux = <5>;
+				function = "gmac1";
+				drive-strength = <40>;
+				bias-disable;
+			};
+
 			/omit-if-no-ref/
 			spdif_out_pb_pin: spdif-pb-pin {
 				pins = "PB8";
@@ -1040,8 +1056,8 @@ gmac0: ethernet@4500000 {
 			compatible = "allwinner,sun55i-a523-gmac0",
 				     "allwinner,sun50i-a64-emac";
 			reg = <0x04500000 0x10000>;
-			clocks = <&ccu CLK_BUS_EMAC0>;
-			clock-names = "stmmaceth";
+			clocks = <&ccu CLK_BUS_EMAC0>, <&ccu CLK_EMAC0_25M>;
+			clock-names = "stmmaceth", "emac25m";
 			resets = <&ccu RST_BUS_EMAC0>;
 			reset-names = "stmmaceth";
 			interrupts = <GIC_SPI 46 IRQ_TYPE_LEVEL_HIGH>;
@@ -1062,8 +1078,9 @@ gmac1: ethernet@4510000 {
 			compatible = "allwinner,sun55i-a523-gmac200",
 				     "snps,dwmac-4.20a";
 			reg = <0x04510000 0x10000>;
-			clocks = <&ccu CLK_BUS_EMAC1>, <&ccu CLK_MBUS_EMAC1>;
-			clock-names = "stmmaceth", "mbus";
+			clocks = <&ccu CLK_BUS_EMAC1>, <&ccu CLK_MBUS_EMAC1>,
+				 <&ccu CLK_EMAC1_25M>;
+			clock-names = "stmmaceth", "mbus", "emac25m";
 			resets = <&ccu RST_BUS_EMAC1>;
 			reset-names = "stmmaceth";
 			interrupts = <GIC_SPI 47 IRQ_TYPE_LEVEL_HIGH>;
-- 
2.54.0

